File: backend/pkg/aura/world/polygons_test.go
package world

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ---- parsing & validation (plan-zone-polygons.md P2) ----------------------

func TestPolygonParses(t *testing.T) {
	const doc = `{
		"name": "Polys",
		"bounds": { "width": 60, "height": 40 },
		"polygons": [
			{ "profile": "Mountains", "blocksMovement": true,
			  "points": [{"x":-10,"y":-10},{"x":10,"y":-10},{"x":10,"y":10},{"x":-10,"y":10}] },
			{ "profile": "Water",
			  "points": [{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}] }
		]
	}`
	z, err := parseZone([]byte(doc))
	require.NoError(t, err)
	require.Len(t, z.Polygons, 2)
	assert.Equal(t, "Mountains", z.Polygons[0].Profile)
	assert.True(t, z.Polygons[0].BlocksMovement)
	assert.Len(t, z.Polygons[0].Points, 4)
	// The zero value is the safe one: a polygon is decorative unless it says so.
	assert.False(t, z.Polygons[1].BlocksMovement)
}

// Every zone shipped before this authors none — the feature is inert at HEAD.
func TestZoneWithoutPolygonsIsValid(t *testing.T) {
	z, err := parseZone([]byte(`{"name":"None","bounds":{"width":60,"height":40}}`))
	require.NoError(t, err)
	assert.Empty(t, z.Polygons)
}

// Each message names the INDEX: a polygon has no id and no unique name, so the
// array position is the only thing an author can search for — the same rule
// regions and paths already follow.
func TestPolygonValidationNamesTheIndex(t *testing.T) {
	cases := []struct {
		name, poly, want string
	}{
		{"empty profile",
			`{"profile":"  ","points":[{"x":0,"y":0},{"x":1,"y":0},{"x":1,"y":1}]}`,
			"polygon 0: profile must not be empty"},
		{"two points are a line, not an area",
			`{"profile":"Water","points":[{"x":0,"y":0},{"x":1,"y":0}]}`,
			"polygon 0: needs at least 3 points to enclose an area, got 2"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			doc := `{"name":"P","bounds":{"width":60,"height":40},"polygons":[` + c.poly + `]}`
			_, err := parseZone([]byte(doc))
			require.Error(t, err)
			assert.EqualError(t, err, c.want)
		})
	}
}

// ⭐ THREE, not two. A polygon encloses an AREA — the region rule, not the path
// rule. Getting this wrong the other way would accept a two-point "area" whose
// collider (P3) has no inside to fill.
func TestThreePointPolygonIsValid(t *testing.T) {
	_, err := parseZone([]byte(`{"name":"P","bounds":{"width":60,"height":40},
		"polygons":[{"profile":"Water","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}]}]}`))
	require.NoError(t, err)
}

// ⭐ A polygon is NOT a region, and this is the pin that says so: they parse
// into different arrays and nothing merges them. Regions answer resolve() for
// footsteps, music and atmosphere; a cave wall must never turn up there.
func TestPolygonsAndRegionsAreSeparateArrays(t *testing.T) {
	z, err := parseZone([]byte(`{"name":"P","bounds":{"width":60,"height":40},
		"regions":[{"profile":"Fields","points":[{"x":0,"y":0},{"x":9,"y":0},{"x":9,"y":9}]}],
		"polygons":[{"profile":"Mountains","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}]}]}`))
	require.NoError(t, err)
	require.Len(t, z.Regions, 1)
	require.Len(t, z.Polygons, 1)
	assert.Equal(t, "Fields", z.Regions[0].Profile)
	assert.Equal(t, "Mountains", z.Polygons[0].Profile)
}

// ⚑ Polygons are COLLISION geometry (P3), so they move with a placed zone's
// origin — unlike regions, which the client offsets itself. A polygon left at
// zone-local coordinates would wall the overworld instead of the cave.
func TestPolygonPointsMoveWithTheZoneOrigin(t *testing.T) {
	z, err := parseZone([]byte(`{"name":"P","bounds":{"width":40,"height":20},
		"origin":{"x":500,"y":300},
		"regions":[{"profile":"Fields","points":[{"x":0,"y":0},{"x":4,"y":0},{"x":4,"y":4}]}],
		"polygons":[{"profile":"Mountains","points":[{"x":0,"y":0},{"x":4,"y":0},{"x":4,"y":4}]}]}`))
	require.NoError(t, err)
	require.NoError(t, Place([]*Zone{z}))

	assert.EqualValues(t, 500, z.Polygons[0].Points[0].X, "polygons are placed")
	assert.EqualValues(t, 300, z.Polygons[0].Points[0].Y)
	assert.EqualValues(t, 0, z.Regions[0].Points[0].X,
		"regions are NOT — the client applies the origin itself, and doing it here too would move them twice")
}

// ---- outlines, on BOTH surface types (plan-zone-polygons.md D3) -----------

func TestOutlineParsesOnBothSurfaceTypes(t *testing.T) {
	z, err := parseZone([]byte(`{"name":"O","bounds":{"width":60,"height":40},
		"paths":[{"profile":"Water","width":3,"points":[{"x":0,"y":0},{"x":5,"y":0}],
		          "outlineProfile":"Coast","outlineWidth":0.5}],
		"polygons":[{"profile":"Water","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}],
		             "outlineProfile":"Ice","outlineWidth":1.25}]}`))
	require.NoError(t, err)
	assert.Equal(t, "Coast", z.Paths[0].OutlineProfile)
	assert.EqualValues(t, 0.5, z.Paths[0].OutlineWidth)
	assert.Equal(t, "Ice", z.Polygons[0].OutlineProfile)
	assert.EqualValues(t, 1.25, z.Polygons[0].OutlineWidth)
}

// Absent-safe: every shape authored before this has neither key.
func TestOutlineIsAbsentSafe(t *testing.T) {
	z, err := parseZone([]byte(`{"name":"O","bounds":{"width":60,"height":40},
		"paths":[{"profile":"Road","width":2,"points":[{"x":0,"y":0},{"x":5,"y":0}]}],
		"polygons":[{"profile":"Water","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}]}]}`))
	require.NoError(t, err)
	assert.Empty(t, z.Paths[0].OutlineProfile)
	assert.EqualValues(t, 0, z.Paths[0].OutlineWidth)
	assert.Empty(t, z.Polygons[0].OutlineProfile)
}

// ⭐ Both HALF-authored forms are refused, and the reason is that both fail
// silently in game: a named profile with no width strokes zero pixels, a width
// with no profile strokes nothing at all. Either one looks exactly like the
// outline feature not working.
func TestHalfAuthoredOutlineIsRefusedOnBothTypes(t *testing.T) {
	cases := []struct{ name, doc, want string }{
		{"path: profile without width",
			`"paths":[{"profile":"Road","width":2,"points":[{"x":0,"y":0},{"x":5,"y":0}],
			  "outlineProfile":"Coast"}]`,
			`path 0: outlineProfile "Coast" needs a positive outlineWidth, got 0`},
		{"path: width without profile",
			`"paths":[{"profile":"Road","width":2,"points":[{"x":0,"y":0},{"x":5,"y":0}],
			  "outlineWidth":0.5}]`,
			"path 0: outlineWidth 0.5 draws nothing without an outlineProfile"},
		{"polygon: profile without width",
			`"polygons":[{"profile":"Water","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}],
			  "outlineProfile":"Ice"}]`,
			`polygon 0: outlineProfile "Ice" needs a positive outlineWidth, got 0`},
		{"polygon: width without profile",
			`"polygons":[{"profile":"Water","points":[{"x":0,"y":0},{"x":5,"y":0},{"x":5,"y":5}],
			  "outlineWidth":2}]`,
			"polygon 0: outlineWidth 2 draws nothing without an outlineProfile"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			_, err := parseZone([]byte(`{"name":"O","bounds":{"width":60,"height":40},` + c.doc + `}`))
			require.Error(t, err)
			assert.EqualError(t, err, c.want)
		})
	}
}

// ⛔ L4 — THE OUTLINE IS DECORATION AND NEVER TOUCHES COLLISION. Otherwise "my
// river blocks wider than I authored it" becomes a debugging session, and the
// outline stops being safely tunable by eye. This is the pin that says so, for
// both builders.
func TestOutlineDoesNotChangeAnyCollider(t *testing.T) {
	bare, err := parseZone([]byte(`{"name":"O","bounds":{"width":200,"height":200},
		"paths":[{"profile":"Water","width":3,"blocksMovement":true,
		          "points":[{"x":-20,"y":0},{"x":20,"y":0}]}],
		"polygons":[{"profile":"Mountains","blocksMovement":true,
		             "points":[{"x":-10,"y":-10},{"x":10,"y":-10},{"x":10,"y":10},{"x":-10,"y":10}]}]}`))
	require.NoError(t, err)
	// The same world with a FAT outline on both — far wider than either shape.
	rimmed, err := parseZone([]byte(`{"name":"O","bounds":{"width":200,"height":200},
		"paths":[{"profile":"Water","width":3,"blocksMovement":true,
		          "points":[{"x":-20,"y":0},{"x":20,"y":0}],
		          "outlineProfile":"Coast","outlineWidth":9}],
		"polygons":[{"profile":"Mountains","blocksMovement":true,
		             "points":[{"x":-10,"y":-10},{"x":10,"y":-10},{"x":10,"y":10},{"x":-10,"y":10}],
		             "outlineProfile":"Ice","outlineWidth":9}]}`))
	require.NoError(t, err)

	assert.Equal(t, PathCorridors(bare), PathCorridors(rimmed))
	bareP, _ := PolygonColliders(bare)
	rimP, _ := PolygonColliders(rimmed)
	assert.Equal(t, bareP, rimP)
}

File: frontend/src/features/polygons/logic/Polygons.test.ts
import {describe, it, expect} from 'vitest';
import {toPolygons} from './Polygons';
import {resolve, loadRegions} from '../../regions/logic/Regions';

// 1 unit = 120 px (api/shared-constants.json pointsPerMeter), the same
// conversion Regions.toRegions and Paths.toPaths apply — pinned here because
// the world and the full-screen map both read it and a disagreement is
// invisible in either alone.
const PX = 120;

const TRI = [{x: 0, y: 0}, {x: 4, y: 0}, {x: 4, y: 4}];

// One outlined polygon, varying only the outline fields.
function outlinedPolygon(over: Record<string, unknown>) {
    return toPolygons([{profile: 'Water', points: TRI, ...over} as never]);
}

describe('toPolygons', () => {
    it('converts server units to world pixels', () => {
        const [poly] = toPolygons([{profile: 'Mountains', points: [{x: -1, y: 0}, {x: 3, y: 2}, {x: 0, y: 5}]}]);
        expect(poly.profile).toBe('Mountains');
        expect(poly.points).toEqual([
            {x: -PX, y: 0}, {x: 3 * PX, y: 2 * PX}, {x: 0, y: 5 * PX},
        ]);
    });

    // Every zone shipped before this authors none.
    it('treats an absent array as no polygons', () => {
        expect(toPolygons(undefined)).toEqual([]);
    });

    // ⭐ THREE points, not two: a polygon encloses an AREA. That is the region
    // rule, and it is the one place a polygon deliberately does NOT follow the
    // path it shares a Tiled layer with.
    it('drops a shape that cannot enclose an area rather than the whole zone', () => {
        const kept = toPolygons([
            {profile: 'Water', points: [{x: 0, y: 0}, {x: 1, y: 0}]},   // a line
            {profile: 'Water', points: [{x: 0, y: 0}]},                  // a point
            {profile: 'Water', points: TRI},                             // fine
        ]);
        expect(kept).toHaveLength(1);
        expect(kept[0].points).toHaveLength(3);
    });

    // ⚑ The origin is applied ONCE, here — the server leaves client-visual
    // geometry zone-local and the client places it (plan-underworld.md U2).
    it('applies a placed zone origin once', () => {
        const [poly] = toPolygons([{profile: 'Water', points: TRI}], {x: 500, y: 300});
        expect(poly.points[0]).toEqual({x: 500 * PX, y: 300 * PX});
        expect(poly.points[1]).toEqual({x: 504 * PX, y: 300 * PX});
    });

    // blocksMovement is read by the SERVER alone. It must not reach the drawn
    // shape at all — a blocking rock and a decorative one look identical.
    it('does not carry blocksMovement into the drawn polygon', () => {
        const [poly] = toPolygons([{profile: 'Water', blocksMovement: true, points: TRI}]);
        expect(poly).not.toHaveProperty('blocksMovement');
    });
});

// ⭐ THE distinction the whole primitive exists for, and the only test that can
// state it: a polygon is not a material. `Regions.resolve()` answers "what is
// underfoot" for footsteps, music and atmosphere, and is heading for
// quest-trigger identity — a cave wall is none of those things.
//
// ⚑ Structurally a Polygon IS a Region (it extends it so the paint code is
// shared), so nothing in the type system stops someone from feeding one array
// into the other. This is what would go red if they did.
describe('polygons are not regions', () => {
    it('a polygon does not answer a resolve() lookup', () => {
        loadRegions([]);
        const inside = {x: 1 * PX, y: 1 * PX};
        // Same shape, same profile, loaded as a REGION: resolve answers.
        loadRegions([{profile: 'Swamp', points: TRI}]);
        expect(resolve('color', inside)).not.toBeUndefined();

        // The polygon array is simply not part of that lookup — there is no
        // `loadPolygons` call resolve() could ever see.
        loadRegions([]);
        toPolygons([{profile: 'Swamp', points: TRI}]);
        expect(resolve('color', inside)).toBe(resolve('color', {x: 9e6, y: 9e6}));
    });
});

// ---- outlines (plan-zone-polygons.md D3) ---------------------------------

describe('polygon outlines', () => {
    it('converts the outline width to world pixels too', () => {
        const [s] = outlinedPolygon({outlineProfile: 'Coast', outlineWidth: 1.5});
        expect(s.outlineProfile).toBe('Coast');
        expect(s.outlineWidth).toBe(1.5 * PX);
    });

    // ⚑ HALF-authored degrades to NO outline, never to half of one. The server
    // refuses both halves, so this is the client's degrade path for a
    // hand-edited file — and either half alone would draw nothing anyway, so
    // the only question is whether the absence is deliberate.
    it('drops a half-authored outline entirely', () => {
        for (const half of [
            {outlineProfile: 'Coast'},
            {outlineWidth: 1.5},
            {outlineProfile: 'Coast', outlineWidth: 0},
            {outlineProfile: 'Coast', outlineWidth: NaN},
            {outlineProfile: '', outlineWidth: 2},
        ]) {
            const [s] = outlinedPolygon(half);
            expect(s).not.toHaveProperty('outlineProfile');
            expect(s).not.toHaveProperty('outlineWidth');
        }
    });

    // The common case: no outline authored at all, and neither key appears.
    it('leaves an un-outlined shape with neither key', () => {
        const [s] = outlinedPolygon({});
        expect(s).not.toHaveProperty('outlineProfile');
        expect(s).not.toHaveProperty('outlineWidth');
    });
});
